File: solve.js
"use strict";

class Stack {
  constructor(items="") {
    this.items = [...items];
  }
  push(item) {
    this.items.push(item);
  }
  pop() {
    return this.items.pop();
  }
  top() {
    return this.items.slice(-1)[0];
  }
  full() {
    return this.items.length === 4;
  }
  empty() {
    return this.items.length === 0;
  }
  missingOne() {
    return this.items.length == 3 && this.monocolor();
  }
  monocolor() {
    return this.items.every(elem => elem === this.items[0]);
  }
  solved() {
    return this.full() && this.monocolor();
  }
  getItems() {
    return this.items;
  }
  toString() {
    return this.items.join("").padEnd(4, " ");
  }
}

class GameState {
  constructor(stackArray) {
    this.stacks = [];
    for (const stack of stackArray) {
      this.stacks.push(new Stack(stack));
    }
  }
  deepcopy() {
    const stacks = [];
    for (const stack of this.stacks) {
      stacks.push(stack.getItems());
    }
    return new GameState(stacks);
  }
  solved() {
    return this.stacks.every(s => s.solved() || s.empty());
  }
  isLegalMove(orig, dest) {
    if (orig === dest) return false;
    if (this.stacks[dest].full()) return false;
    if (this.stacks[orig].empty() || this.stacks[orig].solved()) return false;
    if (this.stacks[dest].empty()) {
      // Not illegal, but moving monocolor stack into empty makes no progress.
      return this.stacks[orig].monocolor() ? false : true;
    }
    // Not illegal, but moving monocolor 3-stack makes no progress.
    if (this.stacks[orig].missingOne()) return false;
    if (this.stacks[orig].top() !== this.stacks[dest].top()) return false;
    return true;
  }
  getLegalMoves() {
    const legal_moves = []; // [from, to]
    for (let orig = 0; orig < this.stacks.length; orig++) {
      for (let dest = 0; dest < this.stacks.length; dest++) {
        if (this.isLegalMove(orig, dest)) {
          legal_moves.push([orig, dest]);
        }
      }
    }
    return legal_moves;
  }
  move(orig, dest) {
    while (this.isLegalMove(orig, dest)) {
      this.stacks[dest].push(this.stacks[orig].pop());
    }
  }
  getKey() {
    return GameState.stacksToKey(this.stacks);
  }
  static stacksToKey(stacks) {
    const key = [];
    for (const stack of stacks) {
      key.push(stack.toString());
    }
    return key.join("|");
  }
  static keyToStacks(key) {
    const stacks= [];
    for (const split of key.split("|")) {
      stacks.push(split.trim());
    }
    return stacks;
  }

  static solve(initial_state) {
    // TODO: Implement a more optimal algorithm than depth-first-search.
    const stack = [initial_state];
    const visited = new Set();
    const parent = new Map();
    while (stack.length > 0) {
      let current_state = stack.pop();
      visited.add(current_state.getKey());
      if (current_state.solved()) {
        let current_key = current_state.getKey();
        const steps = [GameState.keyToStacks(current_key)];
        while (current_key !== initial_state.getKey()) {
          steps.push(GameState.keyToStacks(parent[current_key]));
          current_key = parent[current_key];
        }
        return steps.reverse();
      }
      for (const move of current_state.getLegalMoves()) {
        const new_state = current_state.deepcopy();
        new_state.move(move[0], move[1]);
        if (!visited.has(new_state.getKey())) {
          stack.push(new_state);
          parent[new_state.getKey()] = current_state.getKey();
          visited.add(new_state.getKey());
        }
      }
    }
    return [];
  }
}

File: solve.py
# The original implementation of the water sort solver in Python.
import copy
import time

class Stack():
    def __init__(self, items=""):
        self._stack = list(items)

    def __eq__(self, other):
        if isinstance(other, type(self)):
            return self._stack == other._stack
        return NotImplemented

    def push(self, item):
        self._stack.append(item)

    def pop(self):
        return self._stack.pop()

    def top(self):
        return self._stack[-1]

    def full(self):
        return len(self._stack) == 4

    def empty(self):
        return len(self._stack) == 0

    def monocolor(self):
        return all(x == self._stack[0] for x in self._stack)

    def missing_one(self):
        return len(self._stack) == 3 and self.monocolor()

    def solved(self):
        return self.full() and self.monocolor()

    def __repr__(self):
        return "Stack('{}')".format("".join(self._stack))

    def __str__(self):
        return "".join(self._stack)

class GameState():
    def __init__(self, stacks):
        self._stacks = [Stack(stack.strip()) for stack in stacks]

    def __hash__(self):
        return hash(self.__str__())

    def __eq__(self, other):
        if isinstance(other, type(self)):
            return self._stacks == other._stacks
        return NotImplemented

    def solved(self):
        return all(stack.solved() or stack.empty() for stack in self._stacks)

    def is_legal_move(self, orig, dest):
        if orig == dest:
            return False
        if self._stacks[dest].full():
            return False
        if self._stacks[orig].empty() or self._stacks[orig].solved():
            return False
        if self._stacks[dest].empty():
            # Legal but strictly sub-optimal
            return False if self._stacks[orig].monocolor() else True
        if self._stacks[orig].missing_one():
            # Legal but strictly sub-optimal
            return False
        if self._stacks[orig].top() != self._stacks[dest].top():
            return False
        return True

    def get_legal_moves(self):
        legal_moves = [] # [from, to]
        for orig in range(len(self._stacks)):
            for dest in range(len(self._stacks)):
                if self.is_legal_move(orig, dest):
                    legal_moves.append([orig, dest])
        return legal_moves

    def move(self, orig, dest):
        while self.is_legal_move(orig, dest):
            self._stacks[dest].push(self._stacks[orig].pop())

    def __repr__(self):
        return "GameState('{}')".format(self.arg_string())

    def arg_string(self):
        return "".join(str(stack).ljust(4) for stack in self._stacks)

    def __str__(self):
        return "|".join(str(stack).ljust(4) for stack in self._stacks)

def solve(initial_state):
    stack = [initial_state]
    visited = {initial_state}
    parent = {}
    while stack:
        current_state = stack.pop()
        if current_state.solved():
            steps = [current_state]
            while current_state != initial_state:
                steps.append(parent[current_state])
                current_state = parent[current_state]
            return reversed(steps)

        for move in current_state.get_legal_moves():
            new_state = copy.deepcopy(current_state)
            new_state.move(*move)
            if new_state not in visited:
                stack.append(new_state)
                parent[new_state] = current_state
                visited.add(new_state)
    return None

state = GameState(["FHDB", "CEEE", "GDHD", "AGBF", "FGHA", "AGIE", "BCHD",
                   "CIFI", "CABI", "", ""])

start = time.time()
solution = solve(state)
print("--- %s seconds ---" % (time.time() - start))
for state in solution:
    print(state)

'''
~/water_sort_solver/ (main) $ python solve.py
--- 0.08284831047058105 seconds ---
FHDB|CEEE|GDHD|AGBF|FGHA|AGIE|BCHD|CIFI|CABI|    |
FHDB|CEEE|GDHD|AGBF|FGHA|AGIE|BCHD|CIFI|CAB |    |I
FHDB|CEEE|GDHD|AGBF|FGHA|AGIE|BCHD|CIFI|CA  |B   |I
FHDB|CEEE|GDHD|AGBF|FGHA|AGIE|BCHD|CIF |CA  |B   |II
FHDB|CEEE|GDHD|AGBF|FGH |AGIE|BCHD|CIF |CAA |B   |II
FHDB|CEEE|GDHD|AGB |FGH |AGIE|BCHD|CIFF|CAA |B   |II
FHDB|CEEE|GDHD|AGBB|FGH |AGIE|BCHD|CIFF|CAA |    |II
FHDB|CEEE|GDHD|AGBB|FGH |AGIE|BCHD|CI  |CAA |FF  |II
FHDB|CEEE|GDHD|AGBB|FGH |AGIE|BCHD|CIII|CAA |FF  |
FHDB|CEEE|GDHD|AG  |FGH |AGIE|BCHD|CIII|CAA |FF  |BB
FHD |CEEE|GDHD|AG  |FGH |AGIE|BCHD|CIII|CAA |FF  |BBB
FHDD|CEEE|GDHD|AG  |FGH |AGIE|BCH |CIII|CAA |FF  |BBB
FHDD|CEEE|GDHD|AG  |FG  |AGIE|BCHH|CIII|CAA |FF  |BBB
FHDD|CEEE|GDHD|AGG |F   |AGIE|BCHH|CIII|CAA |FF  |BBB
FHDD|CEEE|GDHD|AGG |FFF |AGIE|BCHH|CIII|CAA |    |BBB
FHDD|CEEE|GDHD|A   |FFF |AGIE|BCHH|CIII|CAA |GG  |BBB
FHDD|CEEE|GDHD|    |FFF |AGIE|BCHH|CIII|CAAA|GG  |BBB
FHDD|CEEE|GDH |D   |FFF |AGIE|BCHH|CIII|CAAA|GG  |BBB
FHDD|CEEE|GDHH|D   |FFF |AGIE|BCH |CIII|CAAA|GG  |BBB
FH  |CEEE|GDHH|DDD |FFF |AGIE|BCH |CIII|CAAA|GG  |BBB
FHHH|CEEE|GD  |DDD |FFF |AGIE|BCH |CIII|CAAA|GG  |BBB
FHHH|CEEE|G   |DDDD|FFF |AGIE|BCH |CIII|CAAA|GG  |BBB
FHHH|CEEE|GGG |DDDD|FFF |AGIE|BCH |CIII|CAAA|    |BBB
FHHH|CEEE|GGG |DDDD|FFF |AGIE|BC  |CIII|CAAA|H   |BBB
F   |CEEE|GGG |DDDD|FFF |AGIE|BC  |CIII|CAAA|HHHH|BBB
    |CEEE|GGG |DDDD|FFFF|AGIE|BC  |CIII|CAAA|HHHH|BBB
AAA |CEEE|GGG |DDDD|FFFF|AGIE|BC  |CIII|C   |HHHH|BBB
AAA |CEEE|GGG |DDDD|FFFF|AGIE|BCC |CIII|    |HHHH|BBB
AAA |CEEE|GGG |DDDD|FFFF|AGIE|BCC |C   |III |HHHH|BBB
AAA |CEEE|GGG |DDDD|FFFF|AGIE|BCCC|    |III |HHHH|BBB
AAA |CEEE|GGG |DDDD|FFFF|AGI |BCCC|E   |III |HHHH|BBB
AAA |CEEE|GGG |DDDD|FFFF|AG  |BCCC|E   |IIII|HHHH|BBB
AAA |CEEE|GGGG|DDDD|FFFF|A   |BCCC|E   |IIII|HHHH|BBB
AAAA|CEEE|GGGG|DDDD|FFFF|    |BCCC|E   |IIII|HHHH|BBB
AAAA|CEEE|GGGG|DDDD|FFFF|CCC |B   |E   |IIII|HHHH|BBB
AAAA|CEEE|GGGG|DDDD|FFFF|CCC |    |E   |IIII|HHHH|BBBB
AAAA|C   |GGGG|DDDD|FFFF|CCC |    |EEEE|IIII|HHHH|BBBB
AAAA|    |GGGG|DDDD|FFFF|CCCC|    |EEEE|IIII|HHHH|BBBB
'''